### PR TITLE
Correct PCT failures after recent security release

### PR DESCRIPTION
**Pull request by Wadeck**
The failures were:

<details>

<summary>CopyArtifactPermissionPropertyTest: testNotWorkWithQueueItemAuthenticator
</summary>

```
[ERROR] testNotWorkWithQueueItemAuthenticator(hudson.plugins.copyartifact.CopyArtifactPermissionPropertyTest)  Time elapsed: 180.051 s  <<< ERROR!
org.junit.runners.model.TestTimedOutException: test timed out after 180 seconds
        at java.lang.Object.wait(Native Method)
        at java.lang.Object.wait(Object.java:502)
        at hudson.remoting.AsyncFutureImpl.get(AsyncFutureImpl.java:75)
        at org.jvnet.hudson.test.JenkinsRule.assertBuildStatus(JenkinsRule.java:1253)
        at hudson.plugins.copyartifact.CopyArtifactPermissionPropertyTest.testNotWorkWithQueueItemAuthenticator(CopyArtifactPermissionPropertyTest.java:348)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
        at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
        at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
        at org.jvnet.hudson.test.JenkinsRule$1.evaluate(JenkinsRule.java:556)
        at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
        at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.lang.Thread.run(Thread.java:748)
```

</details>

<details>

<summary>LegacyJobConfigMigrationMonitorMigrationTest compilation issue</summary>

```
[ERROR] initializationError(hudson.plugins.copyartifact.monitor.LegacyJobConfigMigrationMonitorMigrationTest)  Time elapsed: 0.001 s  <<< ERROR!
java.lang.NoClassDefFoundError: jenkins/scm/impl/mock/AbstractSampleDVCSRepoRule
        at java.lang.ClassLoader.defineClass1(Native Method)
        at java.lang.ClassLoader.defineClass(ClassLoader.java:763)
        at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
        at java.net.URLClassLoader.defineClass(URLClassLoader.java:467)
        at java.net.URLClassLoader.access$100(URLClassLoader.java:73)
        at java.net.URLClassLoader$1.run(URLClassLoader.java:368)
        at java.net.URLClassLoader$1.run(URLClassLoader.java:362)
        at java.security.AccessController.doPrivileged(Native Method)
        at java.net.URLClassLoader.findClass(URLClassLoader.java:361)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
        at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:335)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        at java.lang.Class.getDeclaredFields0(Native Method)
        at java.lang.Class.privateGetDeclaredFields(Class.java:2583)
        at java.lang.Class.getDeclaredFields(Class.java:1916)
        at org.junit.runners.model.TestClass.getSortedDeclaredFields(TestClass.java:77)
        at org.junit.runners.model.TestClass.scanAnnotatedMembers(TestClass.java:70)
        at org.junit.runners.model.TestClass.<init>(TestClass.java:57)
        at org.junit.runners.ParentRunner.createTestClass(ParentRunner.java:88)
        at org.junit.runners.ParentRunner.<init>(ParentRunner.java:83)
        at org.junit.runners.BlockJUnit4ClassRunner.<init>(BlockJUnit4ClassRunner.java:65)
        at org.junit.internal.builders.JUnit4Builder.runnerForClass(JUnit4Builder.java:10)
        at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:59)
        at org.junit.internal.builders.AllDefaultPossibilitiesBuilder.runnerForClass(AllDefaultPossibilitiesBuilder.java:26)
        at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:59)
        at org.junit.internal.requests.ClassRequest.getRunner(ClassRequest.java:33)
        at org.junit.internal.requests.FilterRequest.getRunner(FilterRequest.java:36)
        at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:362)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:290)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
        at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
        at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
        at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)
Caused by: java.lang.ClassNotFoundException: jenkins.scm.impl.mock.AbstractSampleDVCSRepoRule
        at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
        at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:335)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        ... 35 more
```

</details>

<details>

<summary>LegacyMonitorDataTest: buildForCurrentUser_noAccessAndNonExistent</summary>

```
[ERROR] buildForCurrentUser_noAccessAndNonExistent(hudson.plugins.copyartifact.monitor.LegacyMonitorDataTest)  Time elapsed: 2.987 s  <<< FAILURE!
java.lang.AssertionError:
Expected: is <false>
     but: was <true>
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
        at org.junit.Assert.assertThat(Assert.java:956)
        at org.junit.Assert.assertThat(Assert.java:923)
        at hudson.plugins.copyartifact.monitor.LegacyMonitorDataTest.buildForCurrentUser_noAccessAndNonExistent(LegacyMonitorDataTest.java:306)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
        at org.jvnet.hudson.test.JenkinsRule$1.evaluate(JenkinsRule.java:556)
        at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
        at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.lang.Thread.run(Thread.java:748)
```

</details>

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -94,7 +94,7 @@
       <dependency>
         <groupId>org.jenkins-ci.plugins</groupId>
         <artifactId>matrix-auth</artifactId>
-        <version>1.2</version>
+        <version>2.0</version>
```
Comment (daniel-beck) [inline pom.xml:97]: Would go with 2.3, it's a much nicer release.

2.4.2 would be a reasonable choice if we increase the core dependency to 2.138.4 (which we should, for https://www.jenkins.io/security/advisory/2018-12-05/#SECURITY-595 and its potential impact on accessible URLs).

Comment (Wadeck) [reply pom.xml:97]: I just took the oldest "recent enough" to have the failure being solved. Up to the maintainer if they want to use more recent version ;)

**Comment by Wadeck** [comment]
Concerning the build failures:
I got locally exactly the same error without my PR.
And no error if using -Djenkins.version=2.176.3  which is the version used until 13 days ago (see also https://github.com/jenkins-infra/pipeline-library/commit/24fa3bc712f488011c4db57f20df8d7bfafef921#commitcomment-38865433)

To solve this issue, it seems that the plugin has to use the BOM but this requires a bump on the core requirement, that I do not want to require to the maintainer.

**Comment by jglick** [comment]
Build failure in this case is because you must use a 4.x parent POM.

**Comment by Wadeck** [comment]
Re-start the CI build as the recommendedSettings was rollbacked.

**Comment by ikedam** [comment]
It looks better to use plugin-pom 4.x and bom.
I'll work for that in https://issues.jenkins-ci.org/browse/JENKINS-62409

**Comment by ikedam** [comment]
@jglick 

> Build failure in this case is because you must use a 4.x parent POM

Would you let me have more information about this?
It sounds unreasonable that CI fails for all plugins using parent pom < 4.0.

File: pom.xml [hunk 2]
```diff
@@ -103,6 +103,18 @@
         <version>3.0</version>
         <scope>test</scope>
       </dependency>
+      <!-- 
+        To fix an issue in PCT with scm-api, this plugin is included by workflow-support
+        Otherwise LegacyJobConfigMigrationMonitorMigrationTest refuses to compile due to lack of
+        jenkins/scm/impl/mock/AbstractSampleDVCSRepoRule
```
Comment (ikedam) [inline pom.xml:109]: Note: This dependency is required via GitSampleRepoRule in git-plugin.
(Maven might lacks the feature to dependencies between artifacts classified as tests).

**Comment by Wadeck** [comment]
@ikedam feel free to close this PR if you are working on pom 4.0 ;)

**Comment by jglick** [comment]
The `buildPlugin` changes were reverted as a hotfix, but my personal advice remains to not use `recommendedConfigurations`.

File: pom.xml (hunk 1 above)
Comment (ikedam) [reply pom.xml:97]: `matrix-auth` is used only for testing purposes and can be replaced with `MockAuthorizationStrategy`.
I'll do that with updating dependencies for BOM.

**Comment by ikedam** [comment]
I misunderstood that `Unable to create live FilePath for EC2` was caused for the issue of pipeline-library.
I got that the issue was already fixed.

**Comment by ikedam** [comment]
I believe this change is useful and should be merged before updating the parent pom and introducing BOM.

@Wadeck 
Would you tell me how I can test this change in my environment?

* `CopyArtifactPermissionPropertyTest#testNotWorkWithQueueItemAuthenticator` still fails with `TestTimedOutException` with jenkins-2.164.1.
    * Tested with `docker run -it --rm -v maven-repo:/root/.m2 -v "$(pwd):/workspace" -w /workspace maven:3.6.3-jdk-11 mvn "-Djenkins.version=2.164.1" "-Dtest=CopyArtifactPermissionPropertyTest#testNotWorkWithQueueItemAuthenticator" test`
    * But it doesn't fail on ci.jenkins.io and it may be a problem with my environment.
* Running PCT fails with the dependency problems. Let me know appropriate Jenkins war version to test against:
    ```
    docker run --rm -v maven-repo:/root/.m2 -v "$(pwd)/out:/pct/out" -e CHECKOUT_SRC=https://github.com/Wadeck/copyartifact-plugin.git -e VERSION=correct-pct-failures jenkins/pct
    ...
    [WARNING]
    [WARNING] Some problems were encountered while building the effective model for org.jenkins-ci.plugins:copyartifact:hpi:1.45-SNAPSHOT
    [WARNING] 'dependencies.dependency.(groupId:artifactId:type:classifier)' must be unique: org.jenkins-ci.plugins.workflow:workflow-step-api:jar -> duplicate declaration of version 2.22 @ org.jenkins-ci.plugins:copyartifact:${revision}${changelist}, /pct/tmp/work/copyartifact/pom.xml, line 262, column 17
    [WARNING] 'dependencies.dependency.(groupId:artifactId:type:classifier)' must be unique: org.jenkins-ci.plugins.workflow:workflow-api:jar -> version ${workflow-api-plugin.version} vs 2.40 @ org.jenkins-ci.plugins:copyartifact:${revision}${changelist}, /pct/tmp/work/copyartifact/pom.xml, line 277, column 17
    [WARNING] 'dependencies.dependency.(groupId:artifactId:type:classifier)' must be unique: org.jenkins-ci.plugins.workflow:workflow-api:jar -> version ${workflow-api-plugin.version} vs 2.40 @ org.jenkins-ci.plugins:copyartifact:${revision}${changelist}, /pct/tmp/work/copyartifact/pom.xml, line 299, column 17
    [WARNING]
    [WARNING] It is highly recommended to fix these problems because they threaten the stability of your build.
    [WARNING]
    [WARNING] For this reason, future Maven versions might no longer support building such malformed projects.
    [WARNING]
    [INFO]
    [INFO] ----------------< org.jenkins-ci.plugins:copyartifact >-----------------
    [INFO] Building Copy Artifact Plugin 1.45-SNAPSHOT
    [INFO] --------------------------------[ hpi ]---------------------------------
    [INFO] ------------------------------------------------------------------------
    [INFO] BUILD FAILURE
    [INFO] ------------------------------------------------------------------------
    ```

**Comment by ikedam** [comment]
> CopyArtifactPermissionPropertyTest#testNotWorkWithQueueItemAuthenticator still fails with TestTimedOutException with jenkins-2.164.1.

Timeout occured even without specifying Jenkins version.
It’s definitely a problem of my environment.

**Comment by Wadeck** [comment]
@ikedam IIRC I tested it with PCT against 2.222

File: pom.xml (hunk 2 above)
Comment (ikedam) [reply pom.xml:109]: I could not reproduce this issue.
Running tests with `-Djenkins.version=2.222` succeeds even without those lines.

But I believe this fix is required as  `mvn dependency:tree` shows that scm-api:tests is not included dependencies.
I rather believe that this problem should have occurred with usual test execution.

**Comment by ikedam** [comment]
I found the problem was come from my environment. Volume mounts on Windows looks cause enormous performance decrease with Java11.

File: pom.xml (hunk 2 above)
Comment (ikedam) [reply pom.xml:109]: I found what happened.

git-plugin:tests >= 3.1.0 depends on `jenkins.scm.impl.mock.AbstractSampleDVCSRepoRule` in org.jenkins-ci.plugins:scm-api.
git-plugin:tests < 3.1.0 depends on `org.jenkinsci.plugins.workflow.steps.scm.AbstractSampleDVCSRepoRule` in org.jenkins-ci.plugins.workflow:workflow-scm-step.

**Comment by ikedam** [comment]
copyartifact-1.45 is released. I believe it fixes PCT.

**Pull request merged**